test/e2e/controller_test.go

```diff
@@ -0,0 +1,401 @@
+	}
+
+	return logicalcluster.New(workspaceName)
+}
+
+func loadClusterConfig(t *testing.T, clusterName logicalcluster.Name) *rest.Config {
+	t.Helper()
+	restConfig, err := config.GetConfigWithContext("base")
+	if err != nil {
+		t.Fatalf("failed to load *rest.Config: %v", err)
+	}
+	return rest.AddUserAgent(kcpclienthelper.SetCluster(rest.CopyConfig(restConfig), clusterName), t.Name())
+}
+
+func loadClient(t *testing.T, clusterName logicalcluster.Name) client.Client {
+	t.Helper()
+	scheme := runtime.NewScheme()
+	if err := clientgoscheme.AddToScheme(scheme); err != nil {
+		t.Fatalf("failed to add client go to scheme: %v", err)
+	}
+	if err := tenancyv1alpha1.AddToScheme(scheme); err != nil {
+		t.Fatalf("failed to add %q to scheme: %v", tenancyv1alpha1.SchemeGroupVersion, err)
+	}
+	if err := catalogv1alpha1.AddToScheme(scheme); err != nil {
+		t.Fatalf("failed to add %q to scheme: %v", catalogv1alpha1.GroupVersion, err)
+	}
+	if err := apisv1alpha1.AddToScheme(scheme); err != nil {
+		t.Fatalf("failed to add %q to scheme: %v", apisv1alpha1.SchemeGroupVersion, err)
+	}
+	tenancyClient, err := client.New(loadClusterConfig(t, clusterName), client.Options{Scheme: scheme})
+	if err != nil {
+		t.Fatalf("failed to create a client: %v", err)
+	}
+	return tenancyClient
+}
+
+func createWorkspace(t *testing.T, clusterName logicalcluster.Name) client.Client {
+	t.Helper()
+	parent, ok := clusterName.Parent()
+	if !ok {
+		t.Fatalf("cluster %q has no parent", clusterName)
+	}
+	c := loadClient(t, parent)
+	t.Logf("creating workspace %q", clusterName)
+	if err := c.Create(context.TODO(), &tenancyv1alpha1.ClusterWorkspace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: clusterName.Base(),
+		},
+		Spec: tenancyv1alpha1.ClusterWorkspaceSpec{
+			Type: tenancyv1alpha1.ClusterWorkspaceTypeReference{
+				Name: "universal",
+				Path: "root",
+			},
+		},
+	}); err != nil {
+		t.Fatalf("failed to create workspace: %q: %v", clusterName, err)
+	}
+
+	t.Logf("waiting for workspace %q to be ready", clusterName)
+	var workspace tenancyv1alpha1.ClusterWorkspace
+	if err := wait.PollImmediate(100*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+		fetchErr := c.Get(context.TODO(), client.ObjectKey{Name: clusterName.Base()}, &workspace)
+		if fetchErr != nil {
+			t.Logf("failed to get workspace %q: %v", clusterName, err)
+			return false, fetchErr
+		}
+		var reason string
+		if actual, expected := workspace.Status.Phase, tenancyv1alpha1.ClusterWorkspacePhaseReady; actual != expected {
+			reason = fmt.Sprintf("phase is %q, not %q", actual, expected)
+			t.Logf("not done waiting for workspace %q to be ready: %q", clusterName, reason)
+		}
+		return reason == "", nil
+	}); err != nil {
+		t.Fatalf("workspace %q never ready: %v", clusterName, err)
+	}
+
+	return createAPIBinding(t, clusterName)
+}
+
+func createAPIBinding(t *testing.T, workspaceCluster logicalcluster.Name) client.Client {
+	c := loadClient(t, workspaceCluster)
```

Comment:
I'd have both this function and createWorkspace take in a client as a parameter, and construct the client once in the `TestXYZ()` functions as needed.